include/tvm/runtime/container.h

```diff
@@ -0,0 +1,259 @@
+  void assign(std::initializer_list<ElemType> init) {
+    assign(init.begin(), init.end());
+  }
+
+  /*!
+   * \brief Access element at index
+   * \param idx The index of the element.
+   * \return Const reference to ElemType at the index.
+   */
+  const ElemType& operator[](size_t idx) const {
+    size_t size = Self()->size();
+    CHECK_LT(idx, size) << "Index " << idx << " out of bounds " << size << "\n";
+    return *(reinterpret_cast<ElemType*>(AddressOf(idx)));
+  }
+
+  /*!
+   * \brief Access element at index
+   * \param idx The index of the element.
+   * \return Reference to ElemType at the index.
+   */
+  ElemType& operator[](size_t idx) { return this->operator[](idx); }
+
+  /*!
+   * \brief Push a value to the end of the array.
+   *
+   * \param val The value to be added.
+   */
+  void push_back(const ElemType& val) {
+    size_t size = Self()->size();
+    CHECK_LT(size, Self()->capacity());
+    (*this)[size] = val;
+    Self()->set_size(size + 1);
+  }
+
+  /*!
+   * \brief Construct a value in the end of the array.
+   *
+   * \tparam Args Type parameters of the arguments.
+   * \param args Arguments used to construct the new value.
+   */
+  template <typename... Args>
+  void emplace_back(Args&&... args) {
+    size_t size = Self()->size();
+    CHECK_LT(size, Self()->capacity());
+    EmplaceInit(size, std::forward<Args>(args)...);
+    Self()->set_size(size + 1);
+  }
+
+  /*!
+   * \brief Destroy the Inplace Array Base object
+   */
+  ~InplaceArrayBase() {
+    if (!IsPOD()) {
+      size_t size = Self()->size();
+      for (size_t i = 0; i < size; ++i) {
+        ElemType* fp = reinterpret_cast<ElemType*>(AddressOf(i));
+        fp->ElemType::~ElemType();
+      }
+    }
+  }
+
+ protected:
+  /*!
+   * \brief Construct a value in place with the arguments.
+   *
+   * \tparam Args Type parameters of the arguments.
+   * \param idx Index of the element.
+   * \param args Arguments to construct the new value.
+   */
+  template <typename... Args>
+  void EmplaceInit(size_t idx, Args&&... args) {
+    size_t size = Self()->size();
+    CHECK_LT(size, Self()->capacity()) << "InplaceArray out of capacity\n";
+    void* field_ptr = AddressOf(size);
+    new (field_ptr) ElemType(std::forward<Args>(args)...);
+    Self()->set_size(size + 1);
+  }
+
+ private:
+  /*!
+   * \brief If the ElemType is Plain Old Data.
+   */
+  inline bool IsPOD() const {
+    return std::is_standard_layout<ElemType>::value &&
+           std::is_trivial<ElemType>::value;
+  }
+
+  /*!
+   * \brief Return the self object for the array.
+   *
+   * \return Pointer to ArrayType.
+   */
+  inline ArrayType* Self() const {
+    return static_cast<ArrayType*>(const_cast<InplaceArrayBase*>(this));
+  }
+
+  /*!
+   * \brief Return the raw pointer to the element at idx.
+   *
+   * \param idx The index of the element.
+   * \return Raw pointer to the element.
+   */
+  void* AddressOf(size_t idx) const {
+    size_t kDataStart = sizeof(ArrayType);
```

Comment:
I am a bit concerned about the alignment here...Shall we have some alignment tweaks here to make sure the inplace new work properly?

Reply:
The alignment tweak is in `ArrayObjAllocator::New`: `typename std::aligned_union<sizeof(ArrayType), ArrayType, ElemType>::type;`